### ISSUE TITLE
feature request: Robbins-Monro type learning rate decay

### ISSUE THREAD

**Issue by jeffpollock9** [tensorflow/tensorflow#20644]
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: N/A
- **TensorFlow installed from (source or binary)**: N/A
- **TensorFlow version (use command below)**: 1.8
- **Python version**: N/A
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

I was wondering if there is any appetite for a [Robbins-Monro](https://en.wikipedia.org/wiki/Stochastic_approximation#Robbins%E2%80%93Monro_algorithm) type learning rate decay in tensorflow? The decay would be roughly (a more general solution is implemented at the bottom):

```python
decayed_learning_rate = learning_rate * (global_step + 1) ^ (-decay_rate) 
```

As far as I can tell it is not already implemented in tensorflow, which surprised me since I think this is the learning rate decay rate required for theoretical convergence using the Adam optimizer in Section 4 of [the paper](https://arxiv.org/pdf/1412.6980.pdf), which has:

```python
alpha_t = alpha / sqrt(t)
```

which is the same as the first equation with `decay_rate = 0.5`, and I assume they start at `t = 1` while tensorflow starts with `global_step = 0`.

I have an implementation I have been using (mostly copied from the [already implemented ones](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/learning_rate_decay.py)):

```python
def robbins_monro_decay(learning_rate,
                        global_step,
                        decay_steps,
                        decay_rate,
                        staircase=False,
                        name=None):
    """A Robbins-Monro type decay"""

    if global_step is None:
        raise ValueError("global_step is required for robbins_monro_decay.")

    with tf.name_scope(
            name,
            "RobbinsMonroDecay",
            [learning_rate, global_step, decay_steps, decay_rate]) as name:

        learning_rate = tf.convert_to_tensor(
            learning_rate, name="learning_rate")
        dtype = learning_rate.dtype
        decay_steps = tf.cast(decay_steps, dtype)
        decay_rate = tf.cast(decay_rate, dtype)
        global_step = tf.cast(global_step, dtype)
        p = global_step / decay_steps

        if staircase:
            p = tf.floor(p)

        return tf.multiply(
            learning_rate, tf.pow(p + 1, -decay_rate), name=name)
```

I can make a full pull request if that would be useful? I would be more than happy to add some additional documentation/code or change any of the code/naming/whatever.

Thanks!

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by tensorflowbutler** [comment]
It has been 16 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jilljenn** [comment]
Interested to try this out!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 90 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmothukuru** [comment]
@jeffpollock9,
Sorry for the delayed response. Can you please check  [tf.keras.optimizers.schedules.ExponentialDecay](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/ExponentialDecay) is what you are looking for? Thanks!

**Comment by jeffpollock9** [comment]
@rmothukuru thanks for getting back on this. The docs for the keras optimizer have:

``` python
def decayed_learning_rate(step):
  return initial_learning_rate * decay_rate ^ (step / decay_steps)
```
whereas the Robbins-Monro one is:

``` python
decayed_learning_rate = initial_learning_rate * (step + 1) ^ (-decay_rate) 
```

So I don't see how they could be the same.

**Comment by jvishnuvardhan** [comment]
@jeffpollock9 May be it is easy to implement the above  Robbins-Monro  learning rate decay through [tf.keras.callbacks.LearningRateScheduler](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/LearningRateScheduler). You could also implement a custom callback to schedule the learning rate.

As this was an old issue, i guess you might have already implemented Robbins-Monro  learning rate decay method. If you want to contribute, you could raise a PR in TF-addons repo. Thanks!

Please let us know what you think. Thanks!

**Comment by jeffpollock9** [comment]
> @jeffpollock9 May be it is easy to implement the above Robbins-Monro learning rate decay through [tf.keras.callbacks.LearningRateScheduler](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/LearningRateScheduler). You could also implement a custom callback to schedule the learning rate.
> 
> As this was an old issue, i guess you might have already implemented Robbins-Monro learning rate decay method. If you want to contribute, you could raise a PR in TF-addons repo. Thanks!
> 
> Please let us know what you think. Thanks!

Hi @jvishnuvardhan I did indeed implement this learning rate, you can see it if you scroll to the top of the page.

I'm not really interested in this any more though unfortunately, mostly because it's been 3 years since opened this issue. If you would like to take my implementation and contribute it though, I think that'd be great. Thanks.

**Comment by jvishnuvardhan** [comment]
Development of keras moved to separate [keras-team/keras](https://github.com/keras-team/keras/issues) repository. 

I will close the issue here and open this in keras-team/keras repo and let team evaluate this feature. Thanks!
